grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1140,110 +1148,452 @@ class StockMovementService {
-            if (picklistsToImport.any{ it.hasErrors() }) {
+            if (picklistItemsToImport.any{ it.hasErrors() }) {
                 return
             }
 
             PickPageItem pickPageItem = pickPageItems.find {
                 it.requisitionItem?.id == requisitionItemId
             }
 
-            if (pickPageItem?.requisitionItem?.picklistItems?.size()) {
-                picklistService.revertPick(requisitionItemId)
-            }
+            // TODO Decide whether we want to use revert pick or clear picklist (below)
+            //if (pickPageItem?.requisitionItem?.picklistItems?.size()) {
+            //    picklistService.revertPick(requisitionItemId)
+            //}
 
+            // TODO Does the order of operations matter (remove shipment items vs picklist items)
+            // Remove existing shipment items and picklist items
             removeShipmentItemsForModifiedRequisitionItem(pickPageItem.requisitionItem)
 
-            picklistsToImport.each { params ->
-                AvailableItem availableItem = pickPageItem.getAvailableItem(params.binLocation, params.lotNumber)
-                Picklist picklist = stockMovement.requisition?.picklist
-                // find existing picklist to update
-                PicklistItem picklistItem = picklist.picklistItems.find {
-                    it.inventoryItem?.id == availableItem.inventoryItem?.id &&
-                            it.binLocation?.id == availableItem.binLocation?.id
+            // TODO Need to test this thoroughly as it does not feel like a good idea. For example,
+            //  we don't necessarily want to clear picklist if there's no picklist items to import
+            //  so this might need to go inside a check on !picklistItemsToImport.empty
+            clearPicklist(pickPageItem.requisitionItem)
+
+            // Iterate over all of the picklist items to import and attempt to create a picklist item
+            picklistItemsToImport.each { params ->
+
+                InventoryItem inventoryItem = pickPageItem.requisitionItem.product.getInventoryItem(params.lotNumber, params.expirationDate)
+                Location internalLocation = command.location.getInternalLocation(params.binLocation)
+
+                // Get the available item for the provide lot number and bin location
+                AvailableItem availableItem = pickPageItem.getAvailableItem(inventoryItem, internalLocation)
+
+                // If there is an available item with enough we allocate that item to the outbound order
+                if (availableItem) {
+                    Picklist picklist = stockMovement.requisition?.picklist
+                    // find existing picklist to update
+                    PicklistItem picklistItem = picklist.picklistItems.find {
+                        it.inventoryItem?.id == availableItem.inventoryItem?.id &&
+                                it.binLocation?.id == availableItem.binLocation?.id
+                    }
```

Comment:
This picklistItem lookup might not work as you expect it to.
I encountered simillar case for picklist item update.
The issue is that before this lookup we call `clearPicklist(pickPageItem.requisitionItem)`
which will delete `PicklistItems` so in the ned we are always looking for Picklistitems that do not exist anymore.

So we either can remove this lookup line and ignore "updating picklist Items" or rethink the `clearPicklist`.

Currently we don't have a case anywhere where we update existing picklistItem (at least I don't think so).
The way we handle update for picklistItems is we remove them and recreate as a new one.